tests/test_edgeql_json.py

```diff
@@ -1408,12 +1408,11 @@ async def test_edgeql_json_slice_03(self):
             """)
 
     async def test_edgeql_json_bytes_cast_01(self):
-        async with self.assertRaisesRegexTx(
-                edgedb.QueryError, r'cannot cast.*bytes.*to.*json.*'):
-
-            await self.con.execute(r"""
-                SELECT <json>b'foo';
-            """)
+        await self.assert_query_result(
+            r""" SELECT <json>b'foo'; """,
+            ['Zm9v'],
+            ['"Zm9v"'],
+        )
```

Comment:
Please also add tests when a json type is in a shape, e.g.

```
SELECT <json>User {
  name,
  some_bytes := b'aaa'
}
```

and the same query without `<json>` but when compiled in JSON output mode.

Reply:
Casting shapes with `<json>` and `query_json()` don't work with the current changes. I've been tinkering with the compiler for a while trying to find the relative parts. If anyone has any pointers let me know.

Reply:
It's an interesting issue, actually.  See `compile_TypeCast` in `edgeql/compiler/expr.py`.  Specifically the trailing `else:` branch sets up a context to compile the casted expression.  What we are missing here is another context flag that would tell the shape processor in `viewgen.py` that it is inside a JSON cast, and that every shape element value should be cast to `<json>`.  We do this type of recursion for collections, but not shapes, evidently.  This isn't an issue for most types, since PostgreSQL's cast is the same as ours, but `bytes` exposed the bug here.

Reply:
changed